lib/aba/validations.rb

```diff
@@ -21,27 +25,50 @@ def valid?
         validations.each do |type, param|
           case type
           when :presence
-            self.errors << "#{attribute} is empty" if value.nil? || value.to_s.empty?
+            errors << "#{attribute} is empty" if value.nil? || value.to_s.empty?
           when :bsb
             unless((param && value.nil?) || value =~ /^\d{3}-\d{3}$/)
-              self.errors << "#{attribute} format is incorrect"
+              errors << "#{attribute} format is incorrect"
             end
           when :max_length
-            self.errors << "#{attribute} length must not exceed #{param} characters" if value.to_s.length > param
+            errors << "#{attribute} length must not exceed #{param} characters" if value.to_s.length > param
           when :length
-            self.errors << "#{attribute} length must be exactly #{param} characters" if value.to_s.length != param
+            errors << "#{attribute} length must be exactly #{param} characters" if value.to_s.length != param
           when :integer
-            self.errors << "#{attribute} must be an integer" unless value.to_s =~ /\A[+-]?\d+\Z/
+            if param
+              errors << "#{attribute} must be an integer" unless value.to_s =~ /\A[+-]?\d+\Z/
+            else
+              errors << "#{attribute} must be an unsigned integer" unless value.to_s =~ /\A\d+\Z/
+            end
+          when :account_number
+            if value.to_s =~ /\A[0\ \-]+\Z/ || value.to_s !~ /\A[a-z\d\-\ ]{1,9}\Z/
+              errors << "#{attribute} must be a valid account number"
+            end
+          when :becs
+            errors << "#{attribute} must not contain invalid characters" unless value.to_s =~ BECS_PATTERN
+          when :indicator
+            list = INDICATORS.join('\', \'')
+            errors << "#{attribute} must be a one of '#{list}'" unless INDICATORS.include?(value.to_s)
+          when :transaction_code
+            list = TRANSACTION_CODES.join(', ')
+            errors << "#{attribute} must be a one of #{list}" unless TRANSACTION_CODES.include?(value.to_i)
           end
         end
       end
+      self.errors = errors if errors.length
 
       self.errors.empty?
     end
-   
+
+    # Updates errors through valid? call
+    def get_errors
```

Comment:
Not quite. The problem with the current code is that `errors` only populates after calling `valid?`. So calling aba.errors can return an empty array, even when there is an issue. It gets more confusing when you make extra calls to `add_transaction`.

I actually tend to think that validation should happen when the objects are initialised. Maybe I'll play with that too.